### PR TITLE
[OpRefactor][Do Not Merge] Clean up operation.py file

### PR DESCRIPTION
**Pull request by mariaschuld**
Reorganise and improve documentation in the `operation.py` file to make using the classes easier.

Note: code review is best done by looking at the final file. I haven't changed anything besides reorganisation and docstrings. Oh and I added a test for a docstring example.

### CODE REVIEW

File: pennylane/operation.py
Comment (mariaschuld) [inline pennylane/operation.py:23]: @josh146 I think I need help rewriting this (since it contains differentiation info I don't understand), or shall we delete it? The information should be rather provided in the classes and the dev guide?

**Comment by mariaschuld** [comment]
Funny that this PR suddenly wakes up to CodeFactor. I will fix the issues on op-templates-decomposition rather!

I am not sure why the other tests do not run?

@albi3ro and @josh146 I have tagged you because the resulting file shows quite nicely what is still to be done. For example, the section on "Methods used for inverse" in `Operation` can hopefully be much simplified after we have a new inverse system. Or the different parts for differentiation may change. And of course the observable arithmetic will move.

File: doc/releases/changelog-dev.md
Comment (josh146) [inline doc/releases/changelog-dev.md:330]: Suggest making this more descriptive :) Or, potentially even removing it altogether if it is just a reorganization

File: pennylane/operation.py
Comment (josh146) [reply pennylane/operation.py:23]: I think I would delete everything from "Most gates support the analytic method,..." to the end of the subsection 🙂 

The first two paragraphs are still quite useful, as they introduce the concept of 'trainable parameters'

File: pennylane/operation.py
Comment (josh146) [inline pennylane/operation.py:286]:
```suggestion
      is a SciPy COO matrix format.
```

File: pennylane/operation.py
Comment (josh146) [inline pennylane/operation.py:322]: I wonder if it might be confusing that the example uses `Operation` and not `Operator`?

File: pennylane/operation.py
Comment (josh146) [inline pennylane/operation.py:402]: it is really nice to have a proper docstring here!

File: pennylane/operation.py
Comment (josh146) [inline pennylane/operation.py:412]: I am _very_ excited for the upcoming change of removing this check

File: pennylane/operation.py
Comment (josh146) [inline pennylane/operation.py:960]: I would say that this one is more general than just differentiation, it is used in a whole host of applications/algorithms 🤔

Comment (josh146) [reply pennylane/operation.py:960]: In fact, this one feels more like a base representation?

**Review by josh146 — APPROVED** [review]
Thanks Maria, I am generally in favour -- my only concern is merge conflicts and issues with getting `op-refactor` reviewed.

Can we leave this PR open, and not merge it in until _after_ `op-refactor` is merged into master?

**Comment by mariaschuld** [comment]
> Can we leave this PR open, and not merge it in until _after_ `op-refactor` is merged into master?

Good idea!

File: pennylane/operation.py
Comment (mariaschuld) [reply pennylane/operation.py:322]: I explained it in the text!

File: pennylane/operation.py
Comment (mariaschuld) [reply pennylane/operation.py:960]: Remind me, is the generator defined as `G` in :math:`e^G`, or :math:`e^{iG}`? I'll add it to representations, we should also do this in the dev guide!

Comment (josh146) [reply pennylane/operation.py]: It is actually defined as `e^{iGx}`! where `x` corresponds to a single trainable parameter. So it is a representation that not all operators will have, but I still feel a representation nonetheless.

We also plan on adding `Operator.exponent`, which I believe to be the more generalized version `e^G`

**Comment by mariaschuld** [comment]
Yep, this one has too many conflicts again, we need to make a new PR some day and merge in immediately!

**Pull request closed without merging**
